pkg/k8sutil/sort.go

```diff
@@ -0,0 +1,155 @@
+// Copyright 2019 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package k8sutil
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+var versionRegexp = regexp.MustCompile("^(v[1-9][0-9]*)((alpha|beta)([1-9][0-9]*))?$")
+
+// LessVersions is a Less() function for sorting CRD versions in descending
+// order using the algorithm described in:
+// https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definition-versioning/#version-priority
+//
+// LessVersions is intended to be used with sort.Interface in the Less() func:
+//
+// 		type versions struct {
+// 			vers []string
+// 			cmp  func(string, string) bool
+// 		}
+//
+// 		func (vs versions) Len() int { return len(vs.vers) }
+// 		func (vs versions) Less(i, j int) bool { return vs.cmp(vs.vers[i], vs.vers[j]) }
+// 		func (vs versions) Swap(i, j int) { vs.vers[i], vs.vers[j] = vs.vers[j], vs.vers[i] }
+//
+//    func main() {
+//    	vs := versions{[]string{"v1", "v1alphav1"}, LessVersions}
+//    	sort.Sort(vs)
+//    }
+//
+func LessVersions(verA, verB string) bool {
```

Comment:
Could we make use of this helper function: https://github.com/kubernetes/apimachinery/blob/master/pkg/version/helpers.go#L67

Reply:
I was not aware of this function :grin:. This simplifies sorting significantly:

```go
func LessVersions(verA, verB string) bool {
	return version.CompareKubeAwareVersionStrings(verA, verB) > 0
}
```

Considering how easy this is to implement, I don't think `LessVersions` is a particularly "good" library to export. I'm now thinking  of moving `LessVersions` to `internal/util/k8sutil` for internal use only.